client/components/app.jsx

```diff
@@ -0,0 +1,136 @@
+      }
+      else this.calculateGrandTotal(this.state.tipSelected)
+    })
+  }
+
+  handleReset() {
+    this.setState({ checkTotalSet: !this.state.checkTotalSet, tipTotal: null, grandTotal: null, checkTotal: '', tipSelected: null, fateDecides: false })
+  }
+
+  handleTipPercentageClick(e) {
+    if (this.state.fateDecides) {
+      const randomTip = Math.round(Math.random() * (40 - 15) + 15);
+      const percentage = '0.' + randomTip;
+      const tipSelected = percentage.split('.')[1] + '%';
+      this.setState({ tipSelected }, () => this.calculateGrandTotal(percentage))
+    } else {
+      const tipSelected = e.target.value;
+      this.setState({ tipSelected }, () => this.calculateGrandtotal(tipSelected));
+    }
+  }
+
+  letFateDecide(e) {
+    this.setState({ fateDecides: true }, () => this.handleTipPercentageClick())
+  }
+
+  calculateGrandtotal(tip) {
+    const { checkTotal, checkTotalSet, tipPercentage, fateDecides } = this.state;
+    if (!fateDecides) {
+      for (const key in tipPercentage) {
+        if (key === tip) {
+          const tipTotal = (tipPercentage[key] * checkTotal).toFixed(2);
+          const grandTotal = parseFloat(tipTotal) + parseFloat(checkTotal);
+          this.setState({ tipTotal, grandTotal });
+        }
+      }
+    } else {
+      const tipTotal = (parseFloat(tip) * checkTotal).toFixed(2);
+      const grandTotal = (parseFloat(tipTotal) + parseFloat(checkTotal)).toFixed(2);
+      this.setState({ tipTotal, grandTotal });
+      return;
+    }
+  }
+
+  render() {
+    const { tipPercentage, checkTotal, checkTotalSet, tipTotal, grandTotal, tipSelected } = this.state;
+    return (
+      <>
+        <Header title="TIP CALCULATOR" />
+        <main className="calculator-container container">
+          <section className="row">
+            <div className="col-12 d-flex">
+              CHECK TOTAL: {checkTotalSet ? <i style={{ marginLeft: '0.5em' }}>${checkTotal}</i> : null}
+            </div>
+            <div className="col d-flex">
+              {checkTotalSet ? <input disabled value={checkTotal} /> : <input placeholder="$" onChange={this.onTotalCostChange} type="text" name="checkTotal" value={checkTotal} />}
+              {checkTotalSet ? <button onClick={this.handleReset}>RESET</button> : <button onClick={this.onSetClick}>SET</button>}
+            </div>
+          </section>
+          <hr />
+          <section className="row tip-percentages">
+            <div className="col d-flex">
+              SELECT TIP PERCENTAGE: {tipSelected ? <i style={{ marginLeft: '0.5em' }}>{tipSelected}</i> : null}
+            </div>
+          </section>
+          <section className="row">
+            {Object.entries(tipPercentage).map((val, i) => {
+              return (
+                <div className={`col-2 d-flex justify-content-center`} key={i}>
+                  <button onClick={this.handleTipPercentageClick} value={val[0]} className="btn btn-outline-dark tip-btn">{val[0]}</button>
```

Comment:
I have to look somewhere else to find out what `val[0]` is. Try naming this specific element of the array for stronger semantic value.